dig_test.go

```diff
@@ -462,6 +462,103 @@ func TestEndToEndSuccess(t *testing.T) {
 			require.NotNil(t, a, "*A should be part of the container through Ret2->Ret1")
 		}))
 	})
+
+	t.Run("named instances", func(t *testing.T) {
+		c := New()
+		type A struct{ idx int }
+
+		// returns three named instances of A
+		type ret struct {
+			Out
+
+			A1 A `name:"first"`
+			A2 A `name:"second"`
+			A3 A `name:"third"`
+		}
+
+		// requires two specific named instances
+		type param struct {
+			In
+
+			A1 A `name:"first"`
+			A3 A `name:"third"`
+		}
+		require.NoError(t, c.Provide(func() ret {
+			return ret{A1: A{1}, A2: A{2}, A3: A{3}}
+		}), "provide for three named instances should succeed")
+		require.NoError(t, c.Invoke(func(p param) {
+			assert.Equal(t, 1, p.A1.idx)
+			assert.Equal(t, 3, p.A3.idx)
+		}), "invoke should succeed, pulling out two named instances")
+	})
+
+	t.Run("named instances recurse", func(t *testing.T) {
+		c := New()
+		type A struct{ idx int }
+
+		type Ret1 struct {
+			Out
+
+			A1 A `name:"first"`
+		}
+		type Ret2 struct {
+			Ret1
+
+			A2 A `name:"second"`
+		}
+		type param struct {
+			In
+
+			A1 A `name:"first"`  // should come from ret1 through ret2
+			A2 A `name:"second"` // should come from ret2
+		}
+		require.NoError(t, c.Provide(func() Ret2 {
+			return Ret2{
+				Ret1: Ret1{
+					A1: A{1},
+				},
+				A2: A{2},
+			}
+		}))
+		require.NoError(t, c.Invoke(func(p param) {
+			assert.Equal(t, 1, p.A1.idx)
+			assert.Equal(t, 2, p.A2.idx)
```

Comment:
can you add another test where you `Invoke` on a type which depends on named parameters?

```
type fooArgs struct {
  dig.In

  A1 A `name:"first"`
  A2 A `name:"second"`
}

c.Provide(func(args fooArgs) (*foo, error) { .. })

c.Invoke(func(f *foo) {
 ..
})
```